test/e2e/E2ESafeManager.t.sol

```diff
@@ -0,0 +1,750 @@
+    assertEq(safeEngine.tokenCollateral(_cType(), bob), _scenario.lockedCollateral, 'unequal collateral');
+  }
+
+  function test_TransferCollateral_Revert_SafeNotAllowed() public {
+    vm.prank(alice);
+    vm.expectRevert(IODSafeManager.SafeNotAllowed.selector);
+    transferCollateral(aliceProxy, bobSafeId, bob, 1 ether);
+  }
+
+  function test_TransferCollateral_CType(Scenario memory _scenario) public happyPath(_scenario) {
+    vm.startPrank(alice);
+    _token.approve(aliceProxy, type(uint256).max);
+    depositCollatAndGenDebt(_cType(), aliceSafeId, _scenario.lockedCollateral, 0, aliceProxy);
+    modifySAFECollateralization(aliceProxy, aliceSafeId, -int256(_scenario.lockedCollateral), 0);
+    bytes memory payload = abi.encodeWithSelector(
+      basicActions.transferCollateralWithCType.selector,
+      address(safeManager),
+      _cType(),
+      aliceSafeId,
+      alice,
+      _scenario.lockedCollateral
+    );
+    ODProxy(aliceProxy).execute(address(basicActions), payload);
+    vm.stopPrank();
+    assertEq(safeEngine.tokenCollateral(_cType(), alice), _scenario.lockedCollateral);
+  }
+
+  function test_TransferCollateral_CType_SafeNotAllowed(Scenario memory _scenario) public happyPath(_scenario) {
+    vm.startPrank(alice);
+    _token.approve(aliceProxy, type(uint256).max);
+    depositCollatAndGenDebt(_cType(), aliceSafeId, _scenario.lockedCollateral, 0, aliceProxy);
+    modifySAFECollateralization(aliceProxy, aliceSafeId, -int256(_scenario.lockedCollateral), 0);
+    vm.stopPrank();
+    bytes memory payload = abi.encodeWithSelector(
+      basicActions.transferCollateralWithCType.selector,
+      address(safeManager),
+      _cType(),
+      aliceSafeId,
+      bob,
+      _scenario.lockedCollateral
+    );
+    vm.prank(bob);
+    vm.expectRevert(IODSafeManager.SafeNotAllowed.selector);
+    safeManager.transferCollateral(_cType(), aliceSafeId, bob, _scenario.lockedCollateral);
+  }
+}
+
+contract E2ESafeManagerTest_TransferInternalCoins is E2ESafeManagerSetUp {
+  event TransferInternalCoins(address indexed _sender, uint256 indexed _safe, address _dst, uint256 _rad);
+
+  function setUp() public override {
+    super.setUp();
+    bobProxy = deployOrFind(bob);
+    bobSafeId = safeManager.openSAFE(_cType(), bobProxy);
+    bobData = safeManager.safeData(bobSafeId);
+  }
+
+  function test_TransferInternalCoins() public {
+    vm.prank(address(timelockController));
+    safeEngine.createUnbackedDebt(aliceData.safeHandler, aliceData.safeHandler, 1 ether * RAY);
+    vm.prank(alice);
+    vm.expectEmit();
+    emit TransferInternalCoins(aliceProxy, aliceSafeId, bobData.safeHandler, 1 ether);
+    transferInternalCoins(aliceProxy, aliceSafeId, bobData.safeHandler, 1 ether);
+    assertEq(safeEngine.coinBalance(bobData.safeHandler), 1 ether);
+  }
+
+  function test_TransferInternalCoins_Revert_SafeNotAllowerd() public {
+    vm.prank(address(timelockController));
```

Comment:
Later, we will modify to use BasicActions to generate internal balance instead of Timelock